server/interceptor/interceptor_test.go

```diff
@@ -0,0 +1,425 @@
+			wantReadonly: true,
+		}, {
+			req:          &trillian.GetSignedMapRootRequest{MapId: 30},
+			fullMethod:   "/trillian.TrillianMap/GetSignedMapRoot",
+			wantID:       30,
+			wantType:     trillian.TreeType_MAP,
+			wantReadonly: true,
+		},
+		{
+			req:        &trillian.SetMapLeavesRequest{MapId: 30},
+			fullMethod: "/trillian.TrillianMap/SetMapLeaves",
+			wantID:     30,
+			wantType:   trillian.TreeType_MAP,
+		},
+	}
+	for _, test := range tests {
+		info, err := getRPCInfo(test.req, test.fullMethod)
+		if err != nil {
+			t.Errorf("getRPCInfo(%T) return err = %v, want = nil", test.req, err)
+			continue
+		}
+		if got, want := info.doesNotHaveTree, test.wantNoTree; got != want {
+			t.Errorf("%T: info.doesNotHaveTree = %v, want = %v", test.req, got, want)
+		}
+		if got, want := info.treeID, test.wantID; got != want {
+			t.Errorf("%T: info.treeID = %v, want = %v", test.req, got, want)
+		}
+		wantOpts := &trees.GetOpts{TreeType: test.wantType, Readonly: test.wantReadonly}
+		if diff := pretty.Compare(info.opts, wantOpts); diff != "" {
+			t.Errorf("%T: info.opts diff:\n%v", test.req, diff)
+		}
+	}
+}
+
+func TestCombine(t *testing.T) {
+	i1 := &fakeInterceptor{key: "key1", val: "foo"}
+	i2 := &fakeInterceptor{key: "key2", val: "bar"}
+	i3 := &fakeInterceptor{key: "key3", val: "baz"}
+	e1 := &fakeInterceptor{err: errors.New("intercept error")}
+
+	handlerErr := errors.New("handler error")
+
+	tests := []struct {
+		desc         string
+		interceptors []*fakeInterceptor
+		handlerErr   error
+		wantCalled   int
+		wantErr      error
+	}{
+		{
+			desc: "noInterceptors",
+		},
+		{
+			desc:         "single",
+			interceptors: []*fakeInterceptor{i1},
+			wantCalled:   1,
+		},
+		{
+			desc:         "multi1",
+			interceptors: []*fakeInterceptor{i1, i2, i3},
+			wantCalled:   3,
+		},
+		{
+			desc:         "multi2",
+			interceptors: []*fakeInterceptor{i3, i1, i2},
+			wantCalled:   3,
+		},
+		{
+			desc:         "handlerErr",
+			interceptors: []*fakeInterceptor{i1, i2},
+			handlerErr:   handlerErr,
+			wantCalled:   2,
+			wantErr:      handlerErr,
+		},
+		{
+			desc:         "interceptErr",
+			interceptors: []*fakeInterceptor{i1, e1, i2},
+			wantCalled:   2,
+			wantErr:      e1.err,
+		},
+	}
+
+	ctx := context.Background()
+	req := "request"
+	info := &grpc.UnaryServerInfo{}
+	for _, test := range tests {
+		if l := len(test.interceptors); l < test.wantCalled {
+			t.Errorf("%v: len(interceptors) = %v, want >= %v", test.desc, l, test.wantCalled)
+			continue
+		}
+
+		intercepts := []grpc.UnaryServerInterceptor{}
+		for _, i := range test.interceptors {
+			i.calls = 0
+			intercepts = append(intercepts, i.run)
+		}
+		intercept := Combine(intercepts...)
+
+		handler := &fakeHandler{resp: "response", err: test.handlerErr}
+		resp, err := intercept(ctx, req, info, handler.run)
+		if err != test.wantErr {
+			t.Errorf("%v: err = %q, want = %q", test.desc, err, test.wantErr)
+			continue
+		}
+
+		called := 0
```

Comment:
All this machinery feels a bit complicated; could it be simplified by (say) accumulating text into a `got string` and then just comparing against `want = " i1 i2 i3 handler"` ?